From 3df55924d8e13c893efc9a6fda591c2be96dc9ae Mon Sep 17 00:00:00 2001
From: "Radu Pirea (NXP OSS)" <radu-nicolae.pirea@oss.nxp.com>
Date: Thu, 10 Feb 2022 15:37:03 +0200
Subject: [PATCH 29/43] serial: linflexuart: wait for TX fifo to be empty

Wait for all the bytes to be sent on the wire before entering in init
mode. If the fifo is not empty, the bytes present in fifo will be never
sent.

Issue: ALB-8504

Upstream-Status: Pending 

Signed-off-by: Radu Pirea (NXP OSS) <radu-nicolae.pirea@oss.nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/serial/serial_linflexuart.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/serial/serial_linflexuart.c b/drivers/serial/serial_linflexuart.c
index 1f22f28b42..6af9470d54 100644
--- a/drivers/serial/serial_linflexuart.c
+++ b/drivers/serial/serial_linflexuart.c
@@ -25,6 +25,7 @@
 #define UARTCR_PC1			BIT(6)
 #define UARTCR_TFBM			BIT(8)
 #define UARTCR_RFBM			BIT(9)
+#define UARTCR_TFC			GENMASK(15, 13)
 #define UARTSR_DTF			BIT(1)
 #define UARTSR_RFE			BIT(2)
 #define UARTSR_RMB			BIT(9)
@@ -129,6 +130,10 @@ static int _linflex_serial_init(struct linflex_fsl *base, ulong rate)
 {
 	volatile u32 ctrl;
 
+	/* wait the TX fifo to be empty */
+	while (__raw_readl(&base->uartcr) & UARTCR_TFC)
+		;
+
 	/* set the Linflex in master|init mode and activate by-pass filter
 	 * (where supported) */
 	ctrl = LINCR1_MME | LINCR1_INIT;
-- 
2.17.1

